File: src/main/java/com/rednorte/sigle/listas_service/controller/ListaEsperaController.java
package com.rednorte.sigle.listas_service.controller;

import com.rednorte.sigle.listas_service.dto.ListaEsperaDTO;
import com.rednorte.sigle.listas_service.model.EstadoLista;
import com.rednorte.sigle.listas_service.model.ListaEspera;
import com.rednorte.sigle.listas_service.model.Paciente;
import com.rednorte.sigle.listas_service.service.ListaEsperaService;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/listas")
@RequiredArgsConstructor
public class ListaEsperaController {

    private final ListaEsperaService listaService;

    @GetMapping
    public ResponseEntity<List<ListaEsperaDTO>> getAll() {
        return ResponseEntity.ok(listaService.getAll().stream()
                .map(ListaEsperaDTO::from)
                .toList());
    }

    @GetMapping("/paginado")
    public ResponseEntity<Map<String, Object>> getAllPaginado(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        Page<ListaEspera> resultado = listaService.getAllPaginado(PageRequest.of(page, size));
        Map<String, Object> response = new HashMap<>();
        response.put("content", resultado.getContent().stream().map(ListaEsperaDTO::from).toList());
        response.put("totalPages", resultado.getTotalPages());
        response.put("totalElements", resultado.getTotalElements());
        response.put("currentPage", resultado.getNumber());
        return ResponseEntity.ok(response);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ListaEsperaDTO> getById(@PathVariable Long id) {
        return ResponseEntity.ok(ListaEsperaDTO.from(listaService.getById(id)));
    }

    @GetMapping("/paciente/{pacienteId}")
    public ResponseEntity<List<ListaEsperaDTO>> getByPacienteId(@PathVariable Long pacienteId) {
        return ResponseEntity.ok(listaService.getByPacienteId(pacienteId).stream()
                .map(ListaEsperaDTO::from)
                .toList());
    }

    @GetMapping("/paciente/{pacienteId}/paginado")
    public ResponseEntity<Map<String, Object>> getByPacienteIdPaginado(
            @PathVariable Long pacienteId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        Page<ListaEspera> resultado = listaService.getByPacienteIdPaginado(pacienteId, PageRequest.of(page, size));
        Map<String, Object> response = new HashMap<>();
        response.put("content", resultado.getContent().stream().map(ListaEsperaDTO::from).toList());
        response.put("totalPages", resultado.getTotalPages());
        response.put("totalElements", resultado.getTotalElements());
        response.put("currentPage", resultado.getNumber());
        return ResponseEntity.ok(response);
    }

    @GetMapping("/paciente/email/{email}")
    public ResponseEntity<List<ListaEsperaDTO>> getByPacienteEmail(@PathVariable String email) {
        try {
            return ResponseEntity.ok(listaService.getByPacienteEmail(email).stream()
                    .map(ListaEsperaDTO::from)
                    .toList());
        } catch (RuntimeException e) {
            return ResponseEntity.ok(List.of());
        }
    }

    @PostMapping("/registrar")
    public ResponseEntity<ListaEspera> registrarPaciente(@Valid @RequestBody RegistroListaRequest request) {
        ListaEspera nueva = listaService.registrarPacienteEnLista(
                request.getPaciente(),
                request.getEspecialidad(),
                request.getDiagnostico(),
                request.getPerteneceGes());
        return ResponseEntity.ok(nueva);
    }

    @GetMapping("/especialidad/{especialidad}")
    public ResponseEntity<List<ListaEsperaDTO>> obtenerPorEspecialidad(@PathVariable String especialidad) {
        return ResponseEntity.ok(listaService.obtenerListasPorEspecialidad(especialidad).stream()
                .map(ListaEsperaDTO::from)
                .toList());
    }

    @PutMapping("/{id}/estado")
    public ResponseEntity<ListaEsperaDTO> updateEstado(
            @PathVariable Long id,
            @RequestParam EstadoLista estado,
            @RequestParam(required = false) String diagnostico) {
        return ResponseEntity.ok(ListaEsperaDTO.from(listaService.updateEstado(id, estado, diagnostico)));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        listaService.delete(id);
        return ResponseEntity.ok().build();
    }

    @Data
    public static class RegistroListaRequest {
        @NotNull(message = "El paciente es obligatorio")
        @Valid
        private Paciente paciente;

        @NotBlank(message = "La especialidad es obligatoria")
        private String especialidad;

        @NotBlank(message = "El diagnóstico es obligatorio")
        private String diagnostico;

        private Boolean perteneceGes;
    }
}

File: src/controllers/listaEsperaController.js
const service = require('../services/listaEsperaService');

const getAll = async (req, res, next) => {
  try { res.json(await service.getAll()); } catch (e) { next(e); }
};

const getAllPaginado = async (req, res, next) => {
  try {
    const page = parseInt(req.query.page) || 0;
    const size = parseInt(req.query.size) || 10;
    res.json(await service.getAllPaginado(page, size));
  } catch (e) { next(e); }
};

const getById = async (req, res, next) => {
  try { res.json(await service.getById(req.params.id)); } catch (e) { next(e); }
};

const getByPacienteId = async (req, res, next) => {
  try { res.json(await service.getByPacienteId(req.params.pacienteId)); } catch (e) { next(e); }
};

const getByPacienteIdPaginado = async (req, res, next) => {
  try {
    const page = parseInt(req.query.page) || 0;
    const size = parseInt(req.query.size) || 10;
    res.json(await service.getByPacienteIdPaginado(req.params.pacienteId, page, size));
  } catch (e) { next(e); }
};

const getByPacienteEmail = async (req, res, next) => {
  try { res.json(await service.getByPacienteEmail(req.params.email)); } catch (e) { next(e); }
};

const registrar = async (req, res, next) => {
  try {
    const { paciente, especialidad, diagnostico, perteneceGes } = req.body;
    if (!paciente || !especialidad || !diagnostico) {
      return res.status(400).json({ error: 'paciente, especialidad y diagnostico son obligatorios' });
    }
    if (!paciente.rut || !paciente.nombre) {
      return res.status(400).json({ error: 'El paciente debe tener rut y nombre' });
    }
    res.json(await service.registrarPacienteEnLista(paciente, especialidad, diagnostico, perteneceGes));
  } catch (e) { next(e); }
};

const getByEspecialidad = async (req, res, next) => {
  try { res.json(await service.obtenerListasPorEspecialidad(req.params.especialidad)); } catch (e) { next(e); }
};

const updateEstado = async (req, res, next) => {
  try {
    const { estado, diagnostico } = req.query;
    const ESTADOS_VALIDOS = ['ESPERA', 'AGENDADO', 'ATENDIDO', 'CANCELADO'];
    if (!estado || !ESTADOS_VALIDOS.includes(estado)) {
      return res.status(400).json({ error: `Estado debe ser uno de: ${ESTADOS_VALIDOS.join(', ')}` });
    }
    res.json(await service.updateEstado(req.params.id, estado, diagnostico));
  } catch (e) { next(e); }
};

const remove = async (req, res, next) => {
  try { await service.remove(req.params.id); res.json({ message: 'Lista eliminada' }); } catch (e) { next(e); }
};

module.exports = {
  getAll, getAllPaginado, getById, getByPacienteId, getByPacienteIdPaginado,
  getByPacienteEmail, registrar, getByEspecialidad, updateEstado, remove,
};
